### PR TITLE
Int b 21071 name column filtering on partial

### PR DESCRIPTION
**Pull request by ryan-mchugh**
## Summary

Is there anything you would like reviewers to give additional scrutiny?

[this article](tbd) explains more about the approach used.

## Verification Steps for the Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [ ] Have the Agility acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

- [ ] Has the branch been pulled in and checked out?
- [ ] Have the BL acceptance criteria been met for this change?
- [ ] Was the CircleCI build successful?
- [ ] Has the code been reviewed from a standards and best practices point of view?

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

1. Access the
2. Login as a
3.

### Frontend

- [ ] There are no aXe warnings for UI.
- [ ] This works in [Supported Browsers and their phone views](https://transcom.github.io/mymove-docs/docs/adrs/Browser-Support/#minimum-browser-requirements) (Chrome, Firefox, Edge).
- [ ] There are no new console errors in the browser devtools.
- [ ] There are no new console errors in the test output.
- [ ] If this PR adds a new component to Storybook, it ensures the component is fully responsive, OR if it is intentionally not, a wrapping div using the `officeApp` class or custom `min-width` styling is used to hide any states the would not be visible to the user.
- [ ] This change meets the standards for [Section 508 compliance](https://www.ssa.gov/accessibility/andi/help/install.html).

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/getting-started/development/logging).
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/backend/guides/golang-guide#querying-the-database-safely) have been satisfied.

#### Any new migrations/schema changes:

- [ ] Follows our guidelines for [Zero-Downtime Deploys](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#zero-downtime-migrations).
- [ ] Have been communicated to #g-database.
- [ ] Secure migrations have been tested following the instructions in our [docs](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#secure-migrations).

### CODE REVIEW

**Comment by samaysofo** [comment]
@ryan-mchugh is the pr supposed to still be a draft ?  it's seems to have some conflicts too.

**Comment by brianmanley-caci** [comment]
@ryan-mchugh one thing I did notice is that if you enter a first name and then the last name (e.g. 'John Smith') it does not return results, but if you enter last name then first name (e.g. 'Smith John') then it does return results. That may not be how they intend to use the filter anyways since the names do appear as [Last, First] on the page, but it is an observation.

**Comment by ryan-mchugh** [comment]
> one thing I did notice is that if you enter a first name and then the last name (e.g. 'John Smith') it does not return results, but if you enter last name then first name (e.g. 'Smith John') then it does return results. That may not be how they intend to use the filter anyways since the names do appear as [Last, First] on the page, but it is an observation.

Note, it does work if I just search for partials in-line with the formatting (i.e. `th John`, `th, john` , `ohn`)

Your comment does point out a limitation with this yep, but this does seem like the most reasonable approach. The "Duty Status" column is also being updated to check partials in a separate backlog, matching the original text within the column, no permutations with words swapped around on the location name, etc.

Let's say the user enters the First name, a space, then the last name (different than the format displayed to the user on the table - last, first) -- The only solution I can think of is 'splitting' the user input on the space delimiter.
Steven brought up on a previous draft review that I should not remove spaces in the case that the first/last name contains spaces in addition to the concatenated space. (`De Andre Jones`)

____________
#### Potential solutions

Now a possible, but more complex solution, would be to potentially split user input on the strings, then compare "trigrams" between the first and last name individually. We can display any that have an arbitrary percentage of similarity to the user input by word. It does get into the weeds, but it's possible. 


OR 

Run a second `ILIKE` check with the first/last names on the columns swapped, and then compare the input text from the user again (another performance impact). Return matching results for both `ILIKE` checks.

_____________

Considering the ILIKE approach here is the proposed solution on the original ticket, and we agreed to that in sprint planning, I'd push to continue along here, and maybe open an enhancement feature later if seen fit. 

**What do you think @brianmanley-caci and others on the PR?**


**TLDR - other solutions possible, but recommend sticking with this one**

**Comment by ryan-mchugh** [comment]
working on conflicts now

**Comment by brianmanley-caci** [comment]
> #### Potential solutions
> Now a possible, but more complex solution, would be to potentially split user input on the strings, then compare "trigrams" between the first and last name individually. We can display any that have an arbitrary percentage of similarity to the user input by word. It does get into the weeds, but it's possible.
> 
> OR
> 
> Run a second `ILIKE` check with the first/last names on the columns swapped, and then compare the input text from the user again (another performance impact). Return matching results for both `ILIKE` checks.
> 
> Considering the ILIKE approach here is the proposed solution on the original ticket, and we agreed to that in sprint planning, I'd push to continue along here, and maybe open an enhancement feature later if seen fit.
> 
> **What do you think @brianmanley-caci and others on the PR?**
> 
> **TLDR - other solutions possible, but recommend sticking with this one**

@ryan-mchugh I don't disagree that this allows the user to filter. Also I would think most of the time maybe they'd just search by a last name. It seems sufficient for reasonable filtering, so I wouldn't suggest holding it up.

**Comment by samaysofo** [comment]
> > #### Potential solutions
> > Now a possible, but more complex solution, would be to potentially split user input on the strings, then compare "trigrams" between the first and last name individually. We can display any that have an arbitrary percentage of similarity to the user input by word. It does get into the weeds, but it's possible.
> > OR
> > Run a second `ILIKE` check with the first/last names on the columns swapped, and then compare the input text from the user again (another performance impact). Return matching results for both `ILIKE` checks.
> > Considering the ILIKE approach here is the proposed solution on the original ticket, and we agreed to that in sprint planning, I'd push to continue along here, and maybe open an enhancement feature later if seen fit.
> > **What do you think @brianmanley-caci and others on the PR?**
> > **TLDR - other solutions possible, but recommend sticking with this one**
> 
> @ryan-mchugh I don't disagree that this allows the user to filter. Also I would think most of the time maybe they'd just search by a last name. It seems sufficient for reasonable filtering, so I wouldn't suggest holding it up.

@ryan-mchugh i played around with the query and it seems  its because your query is concatenating the last+first then doing the filter, when i added or concatenation for first+last it seems to work fine. you can try it out with this select statement.
```
select*
FROM service_members
WHERE (first_name || ' ' || last_name) || (last_name || ' ' || first_name)  ILIKE ('%Leo Spacemen%')
```

File: src/pages/Office/MoveQueue/MoveQueue.jsx [hunk 1]
```diff
@@ -62,20 +62,13 @@ export const columns = (moveLockFlag, isQueueManagementEnabled, showBranchFilter
...
-        id: 'lastName',
+        id: 'customerName',
         isFilterable: true,
         exportValue: (row) => {
-          return `${row.customer.last_name}, ${row.customer.first_name}`;
+          return row.customer.dodID;
         },
       },
     ),
-    createHeader('DoD ID', 'customer.dodID', {
```
Comment (stevengleason-caci) [inline src/pages/Office/MoveQueue/MoveQueue.jsx:72]: You removed the DoDID column?
And above, the 'customerName' is returning dodID?

File: src/pages/Office/MoveQueue/MoveQueue.jsx [hunk 2]
```diff
@@ -62,20 +62,13 @@ export const columns = (moveLockFlag, isQueueManagementEnabled, showBranchFilter
         );
       },
       {
-        id: 'lastName',
+        id: 'customerName',
         isFilterable: true,
         exportValue: (row) => {
-          return `${row.customer.last_name}, ${row.customer.first_name}`;
+          return row.customer.dodID;
```
Comment (stevengleason-caci) [inline src/pages/Office/MoveQueue/MoveQueue.jsx:68]: return dodID, but the id is 'customerName'?

File: src/pages/Office/MoveQueue/MoveQueue.jsx (hunk 1 above)
Comment (ryan-mchugh) [reply src/pages/Office/MoveQueue/MoveQueue.jsx:72]: Yep, merge conflicts...

**Comment by ryan-mchugh** [comment]
> > > #### Potential solutions
> > > Now a possible, but more complex solution, would be to potentially split user input on the strings, then compare "trigrams" between the first and last name individually. We can display any that have an arbitrary percentage of similarity to the user input by word. It does get into the weeds, but it's possible.
> > > OR
> > > Run a second `ILIKE` check with the first/last names on the columns swapped, and then compare the input text from the user again (another performance impact). Return matching results for both `ILIKE` checks.
> > > Considering the ILIKE approach here is the proposed solution on the original ticket, and we agreed to that in sprint planning, I'd push to continue along here, and maybe open an enhancement feature later if seen fit.
> > > **What do you think @brianmanley-caci and others on the PR?**
> > > **TLDR - other solutions possible, but recommend sticking with this one**
> > 
> > 
> > @ryan-mchugh I don't disagree that this allows the user to filter. Also I would think most of the time maybe they'd just search by a last name. It seems sufficient for reasonable filtering, so I wouldn't suggest holding it up.
> 
> @ryan-mchugh i played around with the query and it seems its because your query is concatenating the last+first then doing the filter, when i added or concatenation for first+last it seems to work fine. you can try it out with this select statement.
> 
> ```
> select*
> FROM service_members
> WHERE (last_name || ' ' || first_name) || (first_name || ' ' || last_name) ILIKE ('%Leo Spacemen%');
> ```

Alrighty, this is pretty much that second option distilled to one where clause instead of an "or". I think I'll go with it, but I do worry about performance issues. We're already skirting the line with one ILIKE query on these columns. If anyone has other suggestions, now is the time.

Will update with that and some new relevant tests.

**Comment by ryan-mchugh** [comment]
> > > #### Potential solutions
> > > Now a possible, but more complex solution, would be to potentially split user input on the strings, then compare "trigrams" between the first and last name individually. We can display any that have an arbitrary percentage of similarity to the user input by word. It does get into the weeds, but it's possible.
> > > OR
> > > Run a second `ILIKE` check with the first/last names on the columns swapped, and then compare the input text from the user again (another performance impact). Return matching results for both `ILIKE` checks.
> > > Considering the ILIKE approach here is the proposed solution on the original ticket, and we agreed to that in sprint planning, I'd push to continue along here, and maybe open an enhancement feature later if seen fit.
> > > **What do you think @brianmanley-caci and others on the PR?**
> > > **TLDR - other solutions possible, but recommend sticking with this one**
> > 
> > 
> > @ryan-mchugh I don't disagree that this allows the user to filter. Also I would think most of the time maybe they'd just search by a last name. It seems sufficient for reasonable filtering, so I wouldn't suggest holding it up.
> 
> @ryan-mchugh i played around with the query and it seems its because your query is concatenating the last+first then doing the filter, when i added or concatenation for first+last it seems to work fine. you can try it out with this select statement.
> 
> ```
> select*
> FROM service_members
> WHERE (last_name || ' ' || first_name) || (first_name || ' ' || last_name) ILIKE ('%Leo Spacemen%');
> ```

Actually - this is great! I misread, concatenating the reverse at the end was not on my radar, but it makes total sense!

File: src/pages/Office/MoveQueue/MoveQueue.jsx (hunk 2 above)
Comment (ryan-mchugh) [reply src/pages/Office/MoveQueue/MoveQueue.jsx]: fixed

File: src/pages/Office/MoveQueue/MoveQueue.jsx (hunk 1 above)
Comment (ryan-mchugh) [reply src/pages/Office/MoveQueue/MoveQueue.jsx]: fixed

**Comment by stevengleason-caci** [comment]
Running make client_test, I'm getting one failure in PaymentRequestQueue.test.jsx on line 188, looking for lastName.

**Comment by stevengleason-caci** [comment]
Looks good.
The unrelated changes match recent pull requests to main which must not have made it into integrationTesting.

I don't seem to have an Approve button though 😕

File: pkg/services/order/order_fetcher.go [hunk 3]
```diff
@@ -512,12 +513,18 @@ func branchFilter(branch *string, needsCounseling bool, ppmCloseoutGblocs bool)
...
+		if name == nil {
+			return
 		}
+
+		// Remove "," that user may enter between names (displayed on frontend column)
+		nameQueryParam := *name
+		removeCharsRegex := regexp.MustCompile("[,]+")
+		nameQueryParam = removeCharsRegex.ReplaceAllString(nameQueryParam, "")
+		nameQueryParam = fmt.Sprintf("%%%s%%", nameQueryParam)
+		query.Where("((service_members.last_name || ' ' || service_members.first_name) || (service_members.first_name || ' ' || service_members.last_name)) ILIKE ?", nameQueryParam)
```
Comment (stevengleason-caci) [inline pkg/services/order/order_fetcher.go:527]: Not necessary, but can this be simplified: 

```suggestion
		query.Where("(service_members.last_name || ' ' || service_members.first_name || ' ' || service_members.last_name) ILIKE ?", nameQueryParam)
```

**Comment by ryan-mchugh** [comment]
@cameroncaci @danieljordan-caci  Just waiting on a review from outside the team.

**Review by danieljordan-caci — APPROVED** [review]
Looks solid & meets AC - send it
